Fix for accessing the strength of field constraints

CodeBardian opened the pull request:
This PR adds a new SIP conversion to enable usage of `fieldConstraintsAndStrength` method when accessing through python bindings. Previously the method would always return `{0: 0}` regardless of which constraints were set.

Fixes #36462

<!--
  BEFORE HITTING SUBMIT -- Please BUILD AND TEST your changes thoroughly. This is YOUR responsibility! Do NOT rely on the QGIS code maintainers to do this for you!!

  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

CodeBardian committed:
fix typo in test

CodeBardian committed:
add SIP conversion for `QMap<QgsFieldConstraints::Constraint, QgsFieldConstraints::ConstraintStrength>`

CodeBardian committed:
add tests for fieldConstraintsAndStrength

troopa81 reviewed (commented):
looks good, some minor comment on the python to cpp part

CodeBardian committed:
add type check and simplify iterating using PyDict_Next

CodeBardian commented on python/core/conversions.sip:
```diff
@@ -2011,6 +2011,118 @@ template<TYPE1, TYPE2>
+    // Get the Python object as an iterator
+    PyObject *iter = PyObject_GetIter(sipPy);
+
+    if (!sipIsErr)
```
oh right, thanks. I didn't actually understand that if `sipIsErr` is null, it means that the code is asking to check the type rather than doing the actual conversion. I have added the type check now.

troopa81 commented on python/core/conversions.sip:
```diff
@@ -2011,6 +2011,95 @@ template<TYPE1, TYPE2>
+
+        while (PyDict_Next(sipPy, &i, &key, &value))
+        {
+            if (!sipCanConvertToEnum(key, sipType_QgsFieldConstraints_Constraint))
```
Actually, mingw build is unhappy and is not aware of sipCanConvertToEnum ... maybe a different sip version. 

Maybe sipCanConvertToType may work ? If not, let's forget about the type check

CodeBardian committed:
replace sipCanConvertToEnum

sipCanConvertToEnum is only available on sip4, instead check whether sipConvertToEnum errors and set a TypeError accordingly

troopa81 commented:
> *** SIP file not up to date: gui/auto_generated/settings/qgssettingseditorwidgetwrapperimpl.sip
 268a269,270

It's not related to your PR. If nobody fixed it in the meantime, I'll do it next week.
